go/store/types/value_store.go

```diff
@@ -563,63 +563,68 @@ func (lvs *ValueStore) GC(ctx context.Context) error {
 		return err
 	}
 	if rootVal == nil {
-		return nil // empty root
+		// empty root
+		return nil
 	}
 
-	keepChunks := make(chan hash.Hash, gcBuffSize)
+	keepChunks := make(chan []hash.Hash, gcBuffSize)
 
-	eg, egCtx := errgroup.WithContext(ctx)
-	eg.Go(func() error { return collector.MarkAndSweepChunks(egCtx, root, keepChunks) })
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return collector.MarkAndSweepChunks(ctx, root, keepChunks)
+	})
 
-	sendHash := func(h hash.Hash) error {
+	keepHashes := func(hs []hash.Hash) error {
 		select {
-		case keepChunks <- h:
+		case keepChunks <- hs:
 			return nil
-		case <-egCtx.Done():
-			return egCtx.Err()
+		case <-ctx.Done():
+			return ctx.Err()
 		}
 	}
-
-	eg.Go(func() error {
-		// send root chunk
-		err := sendHash(root)
-		if err != nil {
-			return err
+	const batchSize = 16384
+	batches := func(hs []hash.Hash) [][]hash.Hash {
+		copied := make([]hash.Hash, len(hs))
+		copy(copied, hs)
+		res := make([][]hash.Hash, 0)
```

Comment:
why make `copied` here? 
also Nit: `var [][]hash.Hash res`

Reply:
I added a comment. It's because the caller goes on to set `toVisit = toVisit[0:0]` and then append things onto `toVisit`. So the caller mutates the parameter slice after the call to `batches`.

https://play.golang.org/p/C1Rhoje_KPo